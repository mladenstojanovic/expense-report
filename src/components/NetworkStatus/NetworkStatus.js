import React from 'react';
import { useSelector } from 'react-redux';
import loader from './loader.svg';
import success from './success.svg';
import {
  NetworkStatusSingleStyle,
  NetworkStatusStyle
} from './NetworkStatus.style';
import { IN_PROGRESS } from '../../store/actions/network/network.constants';

const NetworkStatus = () => {
  const network = useSelector(state => state.network);
  const NOT_STARTED = 'Waiting...';

  return (
    <NetworkStatusStyle>
      <NetworkStatusSingleStyle>
        <img
          src={network.token === IN_PROGRESS ? loader : success}
          alt="loader"
        />{' '}
        <h2>Authenticating: {network.token || NOT_STARTED}</h2>
      </NetworkStatusSingleStyle>
      <NetworkStatusSingleStyle>
        <img
          src={network.token === IN_PROGRESS ? loader : success}
          alt="loader"
        />
        <h2>Creating User: {network.createUser || NOT_STARTED}</h2>
      </NetworkStatusSingleStyle>
      <NetworkStatusSingleStyle>
        <img
          src={network.token === IN_PROGRESS ? loader : success}
          alt="loader"
        />
        <h2>Connecting to the bank: {network.addConnection || NOT_STARTED}</h2>
      </NetworkStatusSingleStyle>
      <NetworkStatusSingleStyle>
        <img
          src={network.token === IN_PROGRESS ? loader : success}
          alt="loader"
        />
        <h2>Getting transactions: {network.getTransactions || NOT_STARTED}</h2>
      </NetworkStatusSingleStyle>
    </NetworkStatusStyle>
  );
};

export default NetworkStatus;
