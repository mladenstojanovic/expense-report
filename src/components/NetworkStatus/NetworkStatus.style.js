import styled from 'styled-components/macro';

export const NetworkStatusStyle = styled.div``;

export const NetworkStatusSingleStyle = styled.div`
  display: flex;
  align-items: center;
`;

export const NetworkStatusErrorStyle = styled.h2`
  color: ${props => props.theme.colors.lightBlue};
`;
