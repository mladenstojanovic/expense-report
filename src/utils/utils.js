import React from 'react';
import { render } from '@testing-library/react';
import { createStore, combineReducers } from 'redux';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import networkReducer from '../store/reducers/network.reducer';
import theme from '../styles/theme';

export const sleep = milliseconds => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const sortTransactionData = data => {
  const sortedData = {};
  for (let index = 0; index < data.length; index++) {
    if (!data[index].subClass) {
      continue;
    }
    if (!sortedData[data[index].subClass.code]) {
      sortedData[data[index].subClass.code] = {
        total: 0,
        numOfTransactions: 0,
        title: data[index].subClass.title
      };
    }
    sortedData[data[index].subClass.code].total += Math.abs(
      parseInt(data[index].amount)
    );
    sortedData[data[index].subClass.code].numOfTransactions++;
  }
  return sortedData;
};

export const isEmailValid = email => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);

export const isPhoneNumberValid = phone =>
  /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/.test(phone);

export const renderWithReduxAndStyles = (
  ui,
  {
    initialState,
    store = createStore(
      combineReducers({ network: networkReducer }),
      initialState
    )
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Provider>
  ),
  store
});
