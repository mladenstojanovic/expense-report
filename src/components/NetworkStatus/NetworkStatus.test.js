import React from 'react';
import NetworkStatus from './NetworkStatus';
import { renderWithReduxAndStyles } from '../../utils/utils';
import {
  IN_PROGRESS,
  SUCCESS
} from '../../store/actions/network/network.constants';

describe('Test NetworkStatus component', () => {
  it('should render correctly when everything in network is loading', () => {
    const { container } = renderWithReduxAndStyles(<NetworkStatus />, {
      initialState: {
        network: {
          token: IN_PROGRESS,
          createUser: IN_PROGRESS,
          addConnection: IN_PROGRESS,
          job: IN_PROGRESS,
          getTransactions: IN_PROGRESS
        }
      }
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly when everything in network is successful', () => {
    const { container } = renderWithReduxAndStyles(<NetworkStatus />, {
      initialState: {
        network: {
          token: SUCCESS,
          createUser: SUCCESS,
          addConnection: SUCCESS,
          job: SUCCESS,
          getTransactions: SUCCESS
        }
      }
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
