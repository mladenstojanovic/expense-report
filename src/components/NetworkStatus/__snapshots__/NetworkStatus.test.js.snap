// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NetworkStatus component should render correctly when everything in network is loading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="loader.svg"
    />
     
    <h2>
      Authenticating
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="loader.svg"
    />
     
    <h2>
      Creating User
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="loader.svg"
    />
     
    <h2>
      Connecting to the bank
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="loader.svg"
    />
     
    <h2>
      Waiting for job to finish
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="loader.svg"
    />
     
    <h2>
      Getting transactions
    </h2>
  </div>
</div>
`;

exports[`Test NetworkStatus component should render correctly when everything in network is successful 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="success.svg"
    />
     
    <h2>
      Authenticating
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="success.svg"
    />
     
    <h2>
      Creating User
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="success.svg"
    />
     
    <h2>
      Connecting to the bank
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="success.svg"
    />
     
    <h2>
      Waiting for job to finish
    </h2>
  </div>
  <div
    class="c0"
  >
    <img
      alt="loader"
      src="success.svg"
    />
     
    <h2>
      Getting transactions
    </h2>
  </div>
</div>
`;
